### PR TITLE
New game database layout

### LINKED ISSUE #53: Rework Game Database

### PR DESCRIPTION
**Pull request by Cari1111**
## Issues

- resolves #53 

## General Changes

- added `game_database_test.py` in `tests`. This test should only enabled for manual testing because it inserts dummy values into the database
- changed file name into `game_database.py`
- changed db from object to module
- added debug log, when new game is inserted

## New Database Layout

- `game_ID`: Is automatically generated (increased) when new row is inserted
- `player1`: ID of user 1
- `player2`: ID of user 2
- `score1`: score of user 1
- `score2`: score of user 2
- `start_time`: start time of the game
- `end_state`: WIN, DRAW, DISCONNECTED
- `winner`: user ID of the winner
- `disconnected`: user ID of the disconected player

## New Methods

I added specific insert methodes which all use the `_insert_game()` methode:

- `insert_won_game` to insert a game that was won
- `insert_disconnected_game` to insert a game where players disconnected
- `insert_drawn_game` to insert a game that was drawn

I added methodes that use the SQLite function `COUNT()` to count things for diffrent statistics:

- `count_played_games`: count all games the user has played (disconnected and drawn games included)
- `count_won_games`: count all games where the user has won
- `count_disconnected_games`: count all games where the user disconnected
- `count_games_with_disconnect`: count all games where the user or the opponent disconnected
- `win_rate`: calculates the win-rate of the user (ignoring disconnected games)

## Still TODO

- [x]  rework naming
- [x] _add count methodes_
- [x]  _ID with _ or not?_: We agreed on `[...]_id`
- [x]  _update docstrings_

### CODE REVIEW

**Comment by PacoSchatz** [comment]
Thanks for your implementation. I had a look at it and tested it locally - it works perfectly. One little thing I found:

in 
https://github.com/martius-lab/teamproject-competition-server/blob/962fc6a294fa104f56bf774592584f603c8c638a/teamprojekt_competition_server/server/databases.py#L70
 you still print the current time, each time an entry is written. Maybe you could convert this to a log message with a few more details, on what exactly was added to the database.

Apart from that it looks good to me. I like, that you added `insert_won_game()` and `insert_disconnected_game()`. That makes it easier to understand what game you are adding. 
But I can't say much, about how often we will use the functions that use the SQLite function `COUNT()` and we will see if any of these are unnecessary or if we might want to add another

File: teamprojekt_competition_server/server/main.py [hunk 1]
```diff
@@ -54,12 +53,29 @@ def main():
...
+        score_other_user=7,
+    )
+    print(gameID1, gameID2, gameID3)
+    print(game_db.get_user_ids(game_id=gameID2))
+    print(game_db.get_game_ids(user_id=23))
+    print(game_db.get_won_game_ids(user_id=23))
+    print(game_db.count_played_games(user_id=23))
+    print(game_db.count_won_games(user_id=23))
+    print(game_db.count_disconnected_games(user_id=23))
+    print(game_db.win_rate(user_id=23))
```
Comment (Raining-Cloud) [inline teamprojekt_competition_server/server/main.py:78]: there is a test folder btw. don't use main for this

File: teamprojekt_competition_server/server/databases.py [hunk 2]
```diff
@@ -1,8 +1,11 @@
 """Databases"""
 
 import sqlite3
+from enum import Enum
+from datetime import datetime
 
 GAME_DB_NAME = "game"
```
Comment (Raining-Cloud) [inline teamprojekt_competition_server/server/databases.py:7]: I am still confused by this, we fix the name of the database but need to create a new object for acsessing the same database everytime ?

**Review by Raining-Cloud — CHANGES_REQUESTED** [review]
I just took a quick look and found some things :)

File: teamprojekt_competition_server/server/databases.py [hunk 3]
```diff
@@ -18,69 +21,291 @@ def __init__(self) -> None:
...
+            int | None: game ID
+        """
+        return self._insert_game(
+            user1_id=user1_id,
+            user2_id=user2_id,
+            score_user_1=score_user_1,
+            score_user_2=score_user_2,
+            start_time=start_time,
+            game_end_state=GameEndState.DRAW.value,
+        )
```
[the remark this answers is not in the document]
Comment (Cari1111) [reply teamprojekt_competition_server/server/databases.py:185]: you can still use the `_insert_game()` methode. I discussed this with @PacoSchatz and we decided that this is more clear to use these wrapper methodes.

File: teamprojekt_competition_server/server/databases.py [hunk 4]
```diff
@@ -18,69 +21,291 @@ def __init__(self) -> None:
...
+            start_time = datetime.now()
+            print(start_time)
+
+        winner_id = None
+        if game_end_state == GameEndState.WIN.value:
+            winner_id = user1_id if is_user1_winner else user2_id
+
+        disconnected_id = None
+        if game_end_state == GameEndState.DISCONNECTED.value:
+            disconnected_id = user1_id if is_user1_disconnected else user2_id
```
[the remark this answers is not in the document]
Comment (Cari1111) [reply teamprojekt_competition_server/server/databases.py:78]: Yes

**Comment by Cari1111** [comment]
@PacoSchatz 
> Thanks for your implementation. I had a look at it and tested it locally - it works perfectly. One little thing I found:
> 
> in
> 
> https://github.com/martius-lab/teamproject-competition-server/blob/962fc6a294fa104f56bf774592584f603c8c638a/teamprojekt_competition_server/server/databases.py#L70
> 
> you still print the current time, each time an entry is written. Maybe you could convert this to a log message with a few more details, on what exactly was added to the database.
> 
> Apart from that it looks good to me. I like, that you added `insert_won_game()` and `insert_disconnected_game()`. That makes it easier to understand what game you are adding. But I can't say much, about how often we will use the functions that use the SQLite function `COUNT()` and we will see if any of these are unnecessary or if we might want to add another

Thanks for the review. I removed the print and added a log message.

File: teamprojekt_competition_server/server/databases.py [hunk 5]
```diff
@@ -18,69 +21,291 @@ def __init__(self) -> None:
...
+                (only relevant for disconnected games)
 
         Returns:
             int: ID of the game
         """
         # TODO separate record file to store game steps
-        self.cursor.execute(f"INSERT INTO game VALUES ({winnerID}, {loserID})")
+        if start_time is None:
+            start_time = datetime.now()
+            print(start_time)
```
[the remark this answers is not in the document]
Comment (Cari1111) [reply teamprojekt_competition_server/server/databases.py]: See response to paco.

File: teamprojekt_competition_server/server/main.py (hunk 1 above)
Comment (Cari1111) [reply teamprojekt_competition_server/server/main.py]: added these tests to a seperate file in this folder

File: teamprojekt_competition_server/server/databases.py (hunk 2 above)
Comment (Cari1111) [reply teamprojekt_competition_server/server/databases.py]: changed it into a module

File: teamprojekt_competition_server/server/databases.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Raining-Cloud) [reply teamprojekt_competition_server/server/databases.py]: Well still unecceserry imo but if all other need this 🤷

**Review by PacoSchatz — APPROVED** [review]

**Review by Raining-Cloud — APPROVED** [review]

**Pull request merged**